File: app/frontend/src/api/glossary.ts
import { apiFetch } from './client'

export interface GlossaryTerm {
  id: number | null
  glossary_id: number
  source_term: string
  target_term: string
  category: string | null
  notes: string | null
  occurrences: number
  is_locked: boolean
}

export interface GlossaryResponse {
  glossary_id: number
  terms: GlossaryTerm[]
}

export async function getGlossary(book_id: number): Promise<GlossaryResponse> {
  const resp = await apiFetch(`/api/v1/books/${book_id}/glossary`, {})
  if (!resp.ok) throw new Error(`glossary fetch failed: ${resp.status}`)
  return resp.json()
}

export async function addTerm(book_id: number, term: Omit<GlossaryTerm, 'id' | 'glossary_id' | 'occurrences'>): Promise<GlossaryTerm> {
  const resp = await apiFetch(`/api/v1/books/${book_id}/glossary/terms`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(term),
  })
  if (!resp.ok) throw new Error(`add term failed: ${resp.status}`)
  return resp.json()
}

export async function updateTerm(book_id: number, term_id: number, fields: Partial<GlossaryTerm>): Promise<GlossaryTerm> {
  const resp = await apiFetch(`/api/v1/books/${book_id}/glossary/terms/${term_id}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(fields),
  })
  if (!resp.ok) throw new Error(`update term failed: ${resp.status}`)
  return resp.json()
}

export async function deleteTerm(book_id: number, term_id: number): Promise<void> {
  const resp = await apiFetch(`/api/v1/books/${book_id}/glossary/terms/${term_id}`, { method: 'DELETE' })
  if (!resp.ok) throw new Error(`delete term failed: ${resp.status}`)
}

export async function autoExtract(book_id: number, source_text: string, translated_text: string): Promise<GlossaryTerm[]> {
  const resp = await apiFetch(`/api/v1/books/${book_id}/glossary/auto-extract`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ source_text, translated_text }),
  })
  if (!resp.ok) throw new Error(`auto-extract failed: ${resp.status}`)
  return resp.json()
}

File: app/backend/app/routers/glossary.py
"""Glossary CRUD endpoints — /api/v1/books/{book_id}/glossary."""
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from ..database import get_session
from ..services.glossary_service import GlossaryService, GlossaryTerm
from ..utils.sanitize import sanitize_text

router = APIRouter(prefix="/books/{book_id}/glossary", tags=["glossary"])


class TermIn(BaseModel):
    source_term: str
    target_term: str
    category: str | None = None
    notes: str | None = None
    is_locked: bool = False


class TermUpdate(BaseModel):
    target_term: str | None = None
    category: str | None = None
    notes: str | None = None
    is_locked: bool | None = None


class GlossaryResponse(BaseModel):
    glossary_id: int
    terms: list[GlossaryTerm]


@router.get("", response_model=GlossaryResponse)
async def list_glossary(
    book_id: int, session: AsyncSession = Depends(get_session),
) -> GlossaryResponse:
    svc = GlossaryService(session)
    g = await svc.get_or_create_for_book(book_id)
    terms = await svc.list_terms(g.id)
    return GlossaryResponse(glossary_id=g.id, terms=terms)


@router.post("/terms", response_model=GlossaryTerm)
async def add_term(
    book_id: int, body: TermIn,
    session: AsyncSession = Depends(get_session),
) -> GlossaryTerm:
    body.source_term = sanitize_text(body.source_term, "source_term")
    body.target_term = sanitize_text(body.target_term, "target_term")
    if body.notes:
        body.notes = sanitize_text(body.notes, "notes")
    svc = GlossaryService(session)
    g = await svc.get_or_create_for_book(book_id)
    return await svc.add_term(
        glossary_id=g.id, source_term=body.source_term, target_term=body.target_term,
        category=body.category, notes=body.notes, is_locked=body.is_locked,
    )


@router.put("/terms/{term_id}", response_model=GlossaryTerm)
async def update_term(
    book_id: int, term_id: int, body: TermUpdate,
    session: AsyncSession = Depends(get_session),
) -> GlossaryTerm:
    if body.target_term is not None:
        body.target_term = sanitize_text(body.target_term, "target_term")
    if body.notes is not None:
        body.notes = sanitize_text(body.notes, "notes")
    svc = GlossaryService(session)
    # V10: Verify term belongs to this book's glossary (BOLA prevention)
    g = await svc.get_or_create_for_book(book_id)
    updated = await svc.update_term(term_id, **body.model_dump(exclude_none=True))
    if updated is None or updated.glossary_id != g.id:
        raise HTTPException(status_code=404, detail="Term not found")
    return updated


@router.delete("/terms/{term_id}")
async def delete_term(
    book_id: int, term_id: int,
    session: AsyncSession = Depends(get_session),
) -> dict:
    svc = GlossaryService(session)
    # V10: Verify term belongs to this book's glossary (BOLA prevention)
    g = await svc.get_or_create_for_book(book_id)
    term = await svc.get_term(term_id)
    if term is None or term.glossary_id != g.id:
        raise HTTPException(status_code=404, detail="Term not found")
    deleted = await svc.delete_term(term_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Term not found")
    return {"deleted": True}


class AutoExtractRequest(BaseModel):
    source_text: str
    translated_text: str


@router.post("/auto-extract", response_model=list[GlossaryTerm])
async def auto_extract(
    book_id: int, body: AutoExtractRequest,
    session: AsyncSession = Depends(get_session),
) -> list[GlossaryTerm]:
    body.source_text = sanitize_text(body.source_text, "source_text")
    body.translated_text = sanitize_text(body.translated_text, "translated_text")
    svc = GlossaryService(session)
    g = await svc.get_or_create_for_book(book_id)
    return await svc.auto_extract_from_translation(
        glossary_id=g.id, source_text=body.source_text, translated_text=body.translated_text,
    )
